x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java

```diff
@@ -206,6 +209,41 @@ public static Expression deepCopyOfField(String name, DataType type) {
      */
     protected abstract Expression build(Source source, List<Expression> args);
 
+    /**
+     * Build an {@link Expression} where all inputs are field references,
+     * <strong>except</strong> those that have been marked with {@link TestCaseSupplier.TypedData#forceLiteral()}.
+     * <p>Test is ignored if the expression is an aggregation.</p>
+     */
+    protected final Expression buildFieldEvaluableExpression(TestCaseSupplier.TestCase testCase) {
+        var expression = resolveSurrogates(buildFieldExpression(testCase));
+
+        assumeFalse("Resolved expression is not an evaluable function", expression instanceof AggregateFunction);
```

Comment:
By having those intermediary methods resolving surrogates and discarding aggregations, we get to apply the same test we apply to normal functions to folded expressions after surrogation

Reply:
Do we need this if we extract bits of `AbstractFunctionTestCase` into a superclass?

Reply:
I think I agree with Nik here, and as I noted down below.  Switching behavior based on class checks here feels like reinventing inheritance, but worse.  I'd much rather pull out the reusable parts of `AbstractFunctionTestCase` into a super class (ideally in a PR that does only that) and then build `AbstractAggregationTestCase` from that.

Reply:
That was my initial idea, but as each case may or may not be surrogated, we would have to duplicate most function tests in the aggregation class. We would still have both test suites for aggs: One if it's an agg, and another if it's a scalar function. And both would still have an `assume`.

E.g. We can't test intermediary states if the agg surrogates to a scalar. And to test the surrogated scalar, we need the original tests.

> Switching behavior based on class checks here feels like reinventing inheritance

I can do something like overriding the `isAggregation()` method in the subclass. However, for everything else (The assumptions mostly), surrogates are purely reflection based, and they can surrogate to both a scalar or another aggregation.

My original approach was about having a base class with the reusable things, and 2 subclasses with the tests (And other specifics). But after remembering the surrogates, I reverted that, as the "scalar specific tests" were required in both.

Would you go with that option, or something else?

Reply:
Summary of the meeting we had: We'll go with having separate classes and separated test cases. If the surrogate ends up in a scalar function, we'll evaluate it, the same way whatever the case we're testing, as the scalar function should already be tested

Reply:
Updated with different base classes for scalars and for aggs